File: 42-trapping-rain-water/trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        ans=0
        i=0
        lmax=0
        rmax=0
        j=len(height)-1
        while i<j:
            lmax=max(lmax,height[i])
            rmax=max(rmax,height[j])
            if lmax<rmax:
                ans+=(lmax-height[i])
                i+=1
            else :
                ans+=(rmax-height[j])
                j-=1
        return ans
        

File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int ans=0;
        int n=height.size();

        int l=0,r=n-1;
        int l_max=0,r_max=0;
        while(l<r){
            l_max=max(l_max,height[l]);
            r_max=max(r_max,height[r]);
            if(l_max<r_max){
                ans+=l_max-height[l];
                l++;
            }else{
                ans+=r_max-height[r];
                r--;
            }
        }
        return ans;
    }
};